### PR TITLE
Precompute radius search and grid sampling on GPU

### PR DESCRIPTION
**Pull request by nicolas-chaulet**
THis PR contains the code for building the grid sampling and neighbours on the CPU during the data loading process. This produces a speed up of 4x on Shapenet training for KPConv.

### CODE REVIEW

File: src/core/data_transform/transforms.py [hunk 1]
```diff
@@ -183,34 +182,36 @@ def new__inc__(key, num_nodes, special_params=None, func=None):
...
+        # Compute sequentially multi_scale indexes on cpu
+        ms_data = MultiScaleData.from_data(data)
+        precomputed = [data]
+        for index in range(self.num_layers):
+            sampler, neighbour_finder = self.strategies["sampler"][index], self.strategies["neighbour_finder"][index]
+            support = precomputed[index]
+            if sampler:
+                query = sampler(support.clone())
+            else:
+                query = support.clone()
```
Comment (tchaton) [inline src/core/data_transform/transforms.py]: Need to check if clone is necessary here.

File: src/datasets/base_dataset.py [hunk 2]
```diff
@@ -153,9 +167,9 @@ def _set_multiscale_transform(self, transform):
                             attr.dataset, "transform", Compose([current_transform, transform]),
                         )
 
-    def set_strategies(self, model, precompute_multi_scale=False):
+    def set_strategies(self, model):
         strategies = model.get_sampling_and_search_strategies()
```
Comment (tchaton) [inline src/datasets/base_dataset.py]: It won't work with model which aren't implementing UNetBaseModel

File: src/datasets/multiscale_data.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
+from typing import List, Optional
+import torch
+import copy
+import torch_geometric
+from torch_geometric.data import Data
+from torch_geometric.data import Batch
+
+
+class MultiScaleData(Data):
```
Comment (tchaton) [inline src/datasets/multiscale_data.py]: I think we should rename to MultiResolutionData, and MultiScale should be if several radiuses are provided.

File: src/datasets/multiscale_data.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    def from_data(cls, data):
+        ms_data = cls()
+        for k, item in data:
+            ms_data[k] = item
+        return ms_data
+
+
+class MultiScaleBatch(MultiScaleData):
+    @staticmethod
+    def from_data_list(data_list, follow_batch=[]):
```
Comment (tchaton) [inline src/datasets/multiscale_data.py]: Great !

File: src/modules/KPConv/blocks.py [hunk 5]
```diff
@@ -47,25 +47,34 @@ def __init__(
         else:
             self.sampler = None
 
-    def forward(self, data):
-        if self.sampler:
-            querry_data = self.sampler(data.clone())
+    def forward(self, data, pre_computed=None, **kwargs):
+        if not hasattr(data, "block_idx"):
+            setattr(data, "block_idx", 0)
```
Comment (tchaton) [inline src/modules/KPConv/blocks.py]: Why do you set setattr(data, "block_idx", 0) ?

**Comment by tchaton** [comment]
Great PR ! Small things to change and should be good to go.

File: src/modules/KPConv/blocks.py (hunk 5 above)
Comment (tchaton) [reply src/modules/KPConv/blocks.py]: Not sure it is good idea. Let's switch with to dictionnary based pre_computed.

File: src/datasets/base_dataset.py (hunk 2 above)
Comment (nicolas-chaulet) [reply src/datasets/base_dataset.py]: No, but I guess that's fine for now

File: src/core/data_transform/transforms.py (hunk 1 above)
Comment (nicolas-chaulet) [reply src/core/data_transform/transforms.py]: Yes it is, otherwise in the forward methods that would use those precomputed data there is a risk that it overrides a previous iteration (for example its features) which would break things

File: src/datasets/multiscale_data.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    for key in follow_batch:
+        batch["{}_batch".format(key)] = []
+
+    cumsum = {key: 0 for key in keys}
+    batch.batch = []
+    for i, data in enumerate(data_list):
+        for key in data.keys:
+            item = data[key]
+            if torch.is_tensor(item) and item.dtype != torch.bool:
+                mask = item >= 0
```
Comment (tchaton) [inline src/datasets/multiscale_data.py]: Do you see any scenario where this mask could be a trouble ?
